lib/Runtime/Library/JavascriptObject.cpp

```diff
@@ -247,6 +247,39 @@ BOOL JavascriptObject::ChangePrototype(RecyclableObject* object, RecyclableObjec
 
     if (isInvalidationOfInlineCacheNeeded)
     {
+        // Invalidate the "instanceof" cache
+        ThreadContext* threadContext = scriptContext->GetThreadContext();
+        threadContext->MapIsInstInlineCaches([threadContext, object](const Js::Var function, Js::IsInstInlineCache* inlineCacheList) {
+            Assert(inlineCacheList != nullptr);
+
+            Js::IsInstInlineCache* curInlineCache;
+            Js::IsInstInlineCache* nextInlineCache;
+            for (curInlineCache = inlineCacheList; curInlineCache != nullptr; curInlineCache = nextInlineCache)
+            {
+                // Stash away the next cache before we potentially zero out current one
+                nextInlineCache = curInlineCache->next;
+
+                bool clearCurrentCache = curInlineCache->type == object->GetType();
+                if (!clearCurrentCache) {
+                    // Check if function prototype contains old prototype
+                    JavascriptOperators::MapObjectAndPrototypes<true>(curInlineCache->type->GetPrototype(), [&](RecyclableObject* obj)
+                        {
+                            if (object->GetType() == obj->GetType())
+                                clearCurrentCache = true;
+                        });
+                }
+
+                if (clearCurrentCache)
+                {
+                    // Fix cache list
+                    // Deletes empty entries
+                    threadContext->UnregisterIsInstInlineCache(curInlineCache, function);
+                    // Actually invalidate current cache
+                    memset(curInlineCache, 0, sizeof(Js::IsInstInlineCache));
```

Comment:
I didn't expect the memset to be needed, did you test without it and have a problem?

Reply:
It seems to not-work without it -> Should be needed

Reply:
https://github.com/chakra-core/ChakraCore/blob/21e41750b607f4a5714d2703fc059bb21ecef34c/lib/Runtime/Base/ThreadContext.cpp#L3891-L3908

Reply:
I thought the UnregisterIsInstInlineCache should remove the reference so the cache could no longer be hit, I only looked at it quickly though and may have been wrong.

If the function has a list of caches and you memset the first one away I fear that you may make the rest of the list inaccessible - which we often will not want to do.

Lists in chakra core are built of nodes, each node contains data and a "next" pointer:
- data (whatever data is needed for this list entry)
- next pointer -> points to the next node in the list (or is a nullptr if this is the end of the list)

If you delete a node (equivalent of memset) then all nodes after it in the list become unreachable.

Reply:
In the function that deletes a cache list it makes sense that every entry is memset away - it's binning the whole list.

If we have to use memset or equivalent here we'll also need to manually move up the next entry - you may alternatively be able to use one of the "List" methods to delete and move up.

Reply:
https://github.com/chakra-core/ChakraCore/blob/21e41750b607f4a5714d2703fc059bb21ecef34c/lib/Runtime/Library/JavascriptObject.cpp#L257-L260

Reply:
That caches a reference to it locally, but it does not update the reference in the list, for that you'd need to update the `->next` property of the previous inlineCache (not the current one).

Reply:
Note, the failing test case is an attempt to read from a cache value that has been memset to nil - so this logic as it stands is not working.

Reply:
I just did a quick test: The `memset` is actually necessary.
Given, that the `ThreadContext::InvalidateIsInstInlineCacheList` function only uses it (no modifications to list), it has to be enough to invalidate a cache (https://github.com/chakra-core/ChakraCore/pull/6858#discussion_r1002564535).

I guess, some code is holding a reference to single cash items, so we have to zero it out.

Reply:
I just did a quick test: Without `memset` it's not working.

>  it does not update the reference in the list

That should be done by `ThreadContext::UnregisterIsInstInlineCache`

> attempt to read from a cache value that has been memset to nil

I'm looking into it

Reply:
The crash actually happens, before my call to `memset`; I didn't notice, that `type` might be null:
https://github.com/chakra-core/ChakraCore/blob/0cfe82d202c0298f4bbde06070f8dbf9099946c8/lib/Runtime/Library/JavascriptFunction.cpp#L3365-L3367

That happens if `instance` is not a `RecyclableObject` like in the failing test:
https://github.com/chakra-core/ChakraCore/blob/0cfe82d202c0298f4bbde06070f8dbf9099946c8/test/Optimizer/test118.js#L79

Reply:
Sorry for guessing wrong about the failed test.